realtek: basic work for RTL8198C (v2)

plappermaul opened the pull request:
These commits add rudimentary support for RTL8198C devices. Integration should attract other developers and supporters and extend visibility of the realtek target.

plappermaul committed:
realtek: redesign SoC detection

SoC detection works fine but the check logic is inverted all the way.
Make it easier to read and prepare for future SoC inclusion.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: detect RTL8198C SoCs

The RTL8198C is based on a dual core MIPS 1074Kc. Add it to the detection
routine.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: add RTL8198c target

add target so we can compile for RTL8198C

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: add LZMA loader

The RTL8198C devices can be booted with an LZMA loader. Provide it
to make smaller images. The loader has been taken and converted from
rampis platform.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: provide lzma loader in build environment

So we can make use of it later on

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: add RTL8198C platform build option

and disable it for all other targets

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

musashino205 commented on target/linux/realtek/image/rtl8198c.mk:
```diff
@@ -0,0 +1,15 @@
+  LOADADDR := 0x80000000
+  LOADER_PLATFORM := rtl8198c
+  LOADER_TYPE := bin
+  LZMA_TEXT_START := 0x84000000
```
```0x82000000``` (32MiB from ```0x80000000```) or ```0x81800000``` (24MiB from ```0x80000000```) may be enough, like ramips target.
https://github.com/openwrt/openwrt/pull/3279

plappermaul replied:
Understood. But will leave it as is because my test kernel + initramfs is quite big for now.

musashino205 commented on target/linux/realtek/dts-5.10/rtl8198c.dtsi:
```diff
@@ -0,0 +1,99 @@
+			reg = <0>;
+		};
+
+		cpu@0 {
```
This shoud be ```cpu@1```.

plappermaul replied:
thx

plappermaul committed:
realtek: new xHCI driver

Initial driver for xHCI controller on RTL8198C SoC.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: add build option for new xHCI driver

Make the new driver available in the kernel.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: provide initial RTL8918C devicetree

With this devicetree to most basic features of the SoC are supported.
Serial console, GIC and SMP works out of the box with no further
patches. USB is enabled with new driver.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul committed:
realtek: add initial F-Secure Sense configuration

Leave the build target commented as we still have a long way
to go. Documentation for interested developers can be found at
https://openwrt.org/toh/f-secure/sense

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul commented:
v2:

- add new xHCI driver
- adapt devicetree

oliv3r commented on target/linux/realtek/Makefile:
```diff
@@ -7,7 +7,7 @@ BOARD:=realtek
 DEVICE_TYPE:=basic
 FEATURES:=ramdisk squashfs
-SUBTARGETS:=rtl838x rtl839x rtl930x rtl931x
+SUBTARGETS:=rtl838x rtl839x rtl930x rtl931x rtl8198c
```
from a sorting point of view, probably keep the new one at the front :)

oliv3r commented on target/linux/realtek/dts-5.10/rtl8198c.dtsi:
```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include <dt-bindings/interrupt-controller/mips-gic.h>
```
interesting, our rtl83xx don't support the ips-gic; and this older one does? curious indeed.

oliv3r commented on target/linux/realtek/dts-5.10/rtl8198c.dtsi:
```diff
@@ -0,0 +1,104 @@
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,38400";
```
any reason to make this the dtsi default?

oliv3r commented on target/linux/realtek/dts-5.10/rtl8198c.dtsi:
```diff
@@ -0,0 +1,104 @@
+		#interrupt-cells = <3>;
+
+		/*
+		 * Declare the interrupt-parent even though the mti,gic
```
mti?

oliv3r commented on target/linux/realtek/dts-5.10/rtl8198c.dtsi:
```diff
@@ -0,0 +1,104 @@
+		interrupt-parent = <&cpuintc>;
+
+		timer {
+			compatible = "mti,gic-timer";
```
guess its a thing :) sorry for the noise

oliv3r commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/prom.c:
```diff
@@ -81,7 +81,7 @@ void __init device_tree_init(void)
 void __init identify_rtl9302(void)
 {
-	switch (sw_r32(RTL93XX_MODEL_NAME_INFO) & 0xfffffff0) {
+	switch (ioread32(RTL93XX_MODEL_NAME_INFO) & 0xfffffff0) {
```
any reason your changing this for rtl93xx?

Neustradamus commented:
@plappermaul: Why have you closed your PR?

aborgda commented:
see it and try t build for rtl8198c
https://github.com/aborgda/openwrt-8198c
